m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+    return "-infinity";
+  }
+
+  // The format of time durations mimics the String method of Go's
+  // time.Duration. It is designed so that the least granular unit is used such
+  // that the first digit is not 0, e.g. 100µs instead of 0.1ms.
+  bool is_negative = bucket_bound.count() < 0;
+  bucket_bound = std::chrono::nanoseconds(std::abs(bucket_bound.count()));
+
+  std::stringstream stream;
+  if (is_negative) {
+    stream << "-";
+  }
+
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(bucket_bound);
+  if (seconds.count() < 1) {
+    // Durations less than one second format use a smaller unit (milli-, micro-,
+    // or nanoseconds) to ensure that the leading digit is non-zero.
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(bucket_bound)
+            .count() > 0) {
+      stream << FormatDuration(bucket_bound, 6);
+      stream << "ms";
+    } else if (std::chrono::duration_cast<std::chrono::milliseconds>(
+                   bucket_bound)
+                   .count() > 0) {
+      stream << FormatDuration(bucket_bound, 3);
+      stream << "µs";
+    } else {
+      stream << bucket_bound.count();
+      stream << "ns";
+    }
+  } else {
+    auto hours = std::chrono::duration_cast<std::chrono::hours>(bucket_bound);
+    auto minutes =
+        std::chrono::duration_cast<std::chrono::minutes>(bucket_bound - hours);
+    auto nanos = bucket_bound - hours - minutes;
+
+    if (hours.count() > 0) {
+      stream << hours.count();
+      stream << "h";
+    }
+
+    if (minutes.count() > 0) {
+      stream << minutes.count();
+      stream << "m";
+    }
+
+    stream << FormatDuration(nanos, 9);
+    stream << "s";
+  }
+
+  return stream.str();
+}
+
+std::string Reporter::Impl::BucketID(uint64_t bucket_id, uint64_t num_buckets) {
+  int len = std::to_string(bucket_id).size();
+
+  // Ensure the ID is at least 4 characters long.
+  len = std::max(len, 4);
+
+  std::string fmt = "%0" + std::to_string(len) + "d";
+
+  std::stringstream stream;
+  stream << boost::format(fmt) % bucket_id;
+
+  return stream.str();
+}
+
+std::string Reporter::Impl::FormatDuration(std::chrono::nanoseconds duration,
+                                           int precision) {
+  // Format the duration into the fraction of the given precision (i.e v /
+  // (10^p)). Omit any trailing zeros and the decimal when there is no
+  // fractional part.
+  double numerator = static_cast<double>(duration.count());
+  double divisor = std::pow(10, precision);
+  std::string str = std::to_string(numerator / divisor);
+
+  // Remove trailing zeros.
+  while (str[str.size() - 1] == '0') {
+    str.erase(str.size() - 1);
+  }
+
+  // Remove the decimal if it is the last character.
+  if (str[str.size() - 1] == '.') {
+    str.erase(str.size() - 1);
+  }
+
+  return str;
+}
+
+void Reporter::Impl::Run() {
+  std::shared_ptr<TCalcTransport> calc_transport(new TCalcTransport());
```

Comment:
maybe include a cursory note that this is guarded by `run_mutex_` too

Reply:
This actually isn't guarded by `run_mutex_`. The scope of the variable is this `Run` method which is single threaded so we don't need a mutex to protect it. Will add a comment to that effect.